pkg/pool/linked_list.go

```diff
@@ -0,0 +1,100 @@
+package pool
+
+import "sync"
+
+func NewDoubleLinkedList[T any]() *DoubleLinkedList[T] {
+	return &DoubleLinkedList[T]{}
+}
+
+type Node[T any] struct {
+	prev  *Node[T]
+	next  *Node[T]
+	value T
+}
+
+func (n *Node[T]) Value() T {
+	return n.value
+}
+
+type DoubleLinkedList[T any] struct {
+	lock sync.Mutex
+	head *Node[T]
+	len  uint64
+}
+
+func (l *DoubleLinkedList[T]) Len() uint64 {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	return l.len
+}
+
+func (l *DoubleLinkedList[T]) Insert(key T) *Node[T] {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	newNode := &Node[T]{
+		value: key,
+		next:  l.head,
+	}
+
+	if l.head != nil {
+		l.head.prev = newNode
+	}
+	l.head = newNode
+
+	l.len++
+
+	return newNode
+}
+
+func (l *DoubleLinkedList[T]) Delete(node *Node[T]) {
```

Comment:
Don't use defer here because you're not returning, manually unlock when you're done